### PR TITLE
Added nf tag and core to stats

### PR DESCRIPTION
**Pull request by kevindweb**
Added nf tag from the recent pr, and the core number to manager stats

## Summary:
Changed and tested the 3 verbosity levels. Spacing should be fine. Since `nf->tag` is no more than 14 characters (excluding `\0`), the beginning has that many spaces reserved. Core number is similar to `sid` and `iid`, so it doesn't change much. 

**Usage:**
Run onvm regularly, and add NFs to see them on the `stdout` output

| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          |  
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        | 👍
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 

**TODO before merging :**
 - [ ] PR is ready for review
 - [X] Maybe wait for shared cpu?

## Review: 
@koolzz 

(optional) **Subscribers:** << @-mention people who probably care about these changes >>

### CODE REVIEW

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Your results will arrive shortly

File: onvm/onvm_mgr/onvm_stats.c [hunk 1]
```diff
@@ -326,12 +326,12 @@ onvm_stats_display_nfs(unsigned difftime, uint8_t verbosity_level) {
         static const char *NF_MSG[3];
 
         NF_MSG[0] =
-            "\nNF IID / SID     rx_pps  /  tx_pps     rx_drop  /  tx_drop         out   /    tonf     /   "
-            "drop\n-----------------------------------------------------------------------------------------------\n";
+            "\nNF TAG         IID / SID / CORE    rx_pps  /  tx_pps     rx_drop  /  tx_drop         out   /    tonf     /   "
+            "drop\n------------------------------------------------------------------------------------------------------------------\n";
         NF_MSG[1] =
-            "\nNF IID / SID     rx_pps  /  tx_pps            rx  /  tx                out   /    tonf     /   drop\n"
+            "\nNF TAG         IID / SID / CORE    rx_pps  /  tx_pps     rx_drop  /  tx_drop         out   /    tonf     /   drop\n"
```
Comment (koolzz) [inline onvm/onvm_mgr/onvm_stats.c]: Define these lines in a header file

**Comment by koolzz** [comment]
Also please include screenshots

**Comment by onvm** [comment]
> In response to PR creation

## CI Message

Run successful see results:
[Results from nimbnode30]
Median TX pps for Speed Tester: 35126881

## Linter Failed

examples/arp_response/arp_response.c:284:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
Total errors found: 1
examples/flow_table/openflow.h:50:  Using deprecated casting style.  Use static_cast<int>(...) instead  [readability/casting] [4]
examples/flow_table/openflow.h:569:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:634:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:771:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:804:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:865:  Extra space before ( in function call  [whitespace/parens] [4]
examples/flow_table/openflow.h:926:  Extra space before ( in function call  [whitespace/parens] [4]
Total errors found: 7
examples/nf_router/nf_router.c:173:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
Total errors found: 1
onvm/onvm_mgr/onvm_pkt.c:68:  Are you taking an address of a cast?  This is dangerous: could be a temp var.  Take the address before doing the cast, rather than after  [runtime/casting] [4]
Total errors found: 1
onvm/onvm_mgr/onvm_stats.c:329:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:330:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:332:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:334:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:414:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
onvm/onvm_mgr/onvm_stats.c:414:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:415:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
onvm/onvm_mgr/onvm_stats.c:415:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:418:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:423:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
onvm/onvm_mgr/onvm_stats.c:423:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:424:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
onvm/onvm_mgr/onvm_stats.c:424:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:427:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
onvm/onvm_mgr/onvm_stats.c:429:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
onvm/onvm_mgr/onvm_stats.c:429:  Lines should be <= 120 characters long  [whitespace/line_length] [5]
Total errors found: 16
onvm/onvm_nflib/onvm_pkt_common.c:98:  Are you taking an address of a cast?  This is dangerous: could be a temp var.  Take the address before doing the cast, rather than after  [runtime/casting] [4]
Total errors found: 1

**Comment by kevindweb** [comment]
Here are the stats output from using `basic_monitor` and `speed_tester` with the new stats. 
Screen Shot 2019-04-22 at 10 25 04 PM
Screen Shot 2019-04-22 at 10 25 37 PM

As you can see the NF tag has to reserve the same size buffer, regardless of the length of the tag

**Comment by koolzz** [comment]
Looks nice, I'll try to review after SD

**Comment by koolzz** [comment]
As we discussed in the meeting:
 - Configure -v stats, shift the 2nd line so it lines up with the first line (I think leaving the fiels below iid/sid/cores is fine rn.
- Test edge cases with max values for these (max length tag, max sid/max iid, etc)
- Code cleanup (I've left some suggestions)

File: onvm/onvm_mgr/onvm_stats.c [hunk 2]
```diff
@@ -408,26 +402,26 @@ onvm_stats_display_nfs(unsigned difftime, uint8_t verbosity_level) {
                 }
 
                 if (verbosity_level == ONVM_RAW_STATS_DUMP) {
-                        fprintf(stats_out, "%s,%u,%u,%" PRIu64 ",%" PRIu64 ",%" PRIu64 ",%" PRIu64 ",%" PRIu64
+                        fprintf(stats_out, "%s,%s,%u,%u,%u,%" PRIu64 ",%" PRIu64 ",%" PRIu64 ",%" PRIu64 ",%" PRIu64
```
Comment (koolzz) [inline onvm/onvm_mgr/onvm_stats.c]: I suggest we define this String the same way we define ONVM_STATS_MSG and others, it takes up like 4 lines here and isn't very readable.

Comment (koolzz) [reply onvm/onvm_mgr/onvm_stats.c]: well same with others obv

**Comment by koolzz** [comment]
Also we'll probably want to wait for the shared cpu changes so we can have 1 nice stats update instead of adding another pr later.

**Comment by kevindweb** [comment]
Here is the updated screenshot of the normal and verbose onvm stats, respectively. 
Screen Shot 2019-04-27 at 4 41 48 PM
Screen Shot 2019-04-27 at 4 42 33 PM

**Comment by koolzz** [comment]
Waiting on merging this because shared cpu has additional stats

**Comment by koolzz** [comment]
@nks5295 brought up a good idea, we can add the number of children into the `-v` stats

**Comment by koolzz** [comment]
Check cleanup in `onvm_nflib_stop` when function isn't completely initialized.

**Comment by kevindweb** [comment]
@onvm testy test

**Comment by onvm** [comment]
> @onvm testy test

## CI Message

Your results will arrive shortly

File: onvm/onvm_mgr/onvm_stats.h [hunk 3]
```diff
@@ -57,6 +57,15 @@
 #define ONVM_STR_STATS_STDERR "stderr"
 #define ONVM_STR_STATS_WEB "web"
 
+extern const char *NF_MSG[3];
+#define ONVM_STATS_MSG "\nNF TAG         IID / SID / CORE    rx_pps  /  tx_pps        rx_drop  /  tx_drop"\
```
Comment (koolzz) [inline onvm/onvm_mgr/onvm_stats.h]: Lets try 
```
#define ONVM_STATS_MSG \
    "\nNF TAG         IID / SID / CORE    rx_pps  /  tx_pps        rx_drop  /  tx_drop         out   /    tonf     /   drop\n"\
...
```
This way its cleaner to read. Also I feel like its fine if it goes over the char limit for these definitions

Comment (kevindweb) [reply onvm/onvm_mgr/onvm_stats.h]: Ok sounds good, and line limits are going to be the only lint errors in this pr which is fine I guess

**Comment by kevindweb** [comment]
Here's the updated screenshots from the different verbosity levels and shared cpu mode. 

For regular stdout, no special arguments: 
Screen Shot 2019-05-26 at 6 41 16 PM
`./go.sh 0,1,2,3 3 0xF0 -a 0x7f000000000 -s stdout`

For shared cpu mode enabled (`./go.sh 0,1,2,3 3 0xF0 -a 0x7f000000000 -s stdout -c`):
Screen Shot 2019-05-26 at 6 40 32 PM

For verbose and shared cpu (`./go.sh 0,1,2,3 3 0xF0 -a 0x7f000000000 -s stdout -c -v`): 
Screen Shot 2019-05-26 at 6 39 35 PM

**Comment by koolzz** [comment]
Instead of working on the bottom in verbose mode can we have 
```
Children / parent / State
5.             / 3.        / W
```
Not sure how to fit the names but you get the idea

File: onvm/onvm_mgr/onvm_stats.c [hunk 4]
```diff
@@ -401,23 +401,17 @@ onvm_stats_display_nfs(unsigned difftime, uint8_t verbosity_level) {
...
-                    "\n---------------------------------------------------------------------------------------------------"
-                    "\n";
-        } else {
-                NF_MSG[1] =
-                    "\nNF IID / SID     rx_pps  /  tx_pps            rx  /  tx                out   /    tonf     /   drop\n"
-                    "               drop_pps  /  drop_pps     rx_drop  /  tx_drop           next  /    buf      /   ret\n"
-                    "\n---------------------------------------------------------------------------------------------------"
-                    "\n";
-        }
+                    "\nNF TAG         IID / SID / CORE    rx_pps  /  tx_pps        rx_drop  /  tx_drop"
```
Comment (koolzz) [inline onvm/onvm_mgr/onvm_stats.c]: Move to the .h file

File: onvm/onvm_mgr/onvm_stats.c [hunk 5]
```diff
@@ -506,30 +500,30 @@ onvm_stats_display_nfs(unsigned difftime, uint8_t verbosity_level) {
...
                                            ",%" PRIu64 ",%" PRIu64 ",%" PRIu64 ",%" PRIu64 ",%" PRIu64 ", %d\n",
-                                buffer, nfs[i].info->instance_id, nfs[i].info->service_id, rx, tx, rx_pps, tx_pps,
-                                rx_drop, tx_drop, rx_drop_rate, tx_drop_rate, act_out, act_tonf, act_drop, act_next,
-                                act_buffer, act_returned, num_wakeups, wakeup_rate, active);
+                                buffer, nfs[i].info->tag, nfs[i].info->instance_id, nfs[i].info->service_id, nfs[i].info->core,
+                                rx, tx, rx_pps, tx_pps, rx_drop, tx_drop, rx_drop_rate, tx_drop_rate, act_out, act_tonf, act_drop,
+                                act_next, act_buffer, act_returned, num_wakeups, wakeup_rate, active);
                 } else if (verbosity_level == 2) {
-                        fprintf(stats_out, "NF  %2u / %-2u  - %9" PRIu64 " / %-9" PRIu64 "  %11" PRIu64 " / %-11" PRIu64
+                        fprintf(stats_out, "%-14s %2u  /  %-2u / %2u    %9" PRIu64 " / %-9" PRIu64 "   %11" PRIu64 " / %-11" PRIu64
```
Comment (koolzz) [inline onvm/onvm_mgr/onvm_stats.c]: Try defining ""%-14s %2u  /  %-2u / %2u    %9" PRIu64 " / %-9" PRIu64 "   %11" PRIu64 " / %-11" PRIu64......."
in the .h file.  Might make this a bit cleaner.

Comment (koolzz) [reply onvm/onvm_mgr/onvm_stats.c]: If you think that moving that into the .h file is harder to maintain you can keep it. You call

**Comment by kevindweb** [comment]
I'll fix the newest change request tomorrow, but here's the updated shared cpu verbose mode:
Screen Shot 2019-05-26 at 9 43 27 PM

**Comment by kevindweb** [comment]
@onvm tell me how it is

**Comment by onvm** [comment]
> @onvm tell me how it is

## CI Message

Your results will arrive shortly

**Comment by koolzz** [comment]
Should we shorten the lines so its a bit better alligned?
```
 IID / SID / CORE
 PNT / S|W / CHLD
```
I nknow its less intuitive but this was the panel looks nice

**Comment by koolzz** [comment]
I think we can do the above comment^ + basically only display the 3rd line in verbose mode (it'll only have shared cpu stats) when shared CPU is enabled.

**Comment by kevindweb** [comment]
Here is the updated picture of verbose and shared cpu
Screen Shot 2019-05-27 at 2 35 48 PM

**Comment by kevindweb** [comment]
@onvm how do you like me now?

**Comment by onvm** [comment]
> @onvm how do you like me now?

## CI Message

Your results will arrive shortly

**Comment by kevindweb** [comment]
@onvm are the updates working?

**Comment by onvm** [comment]
> @onvm are the updates working?

## CI Message

Your results will arrive shortly

File: onvm/onvm_mgr/onvm_stats.h [hunk 6]
```diff
@@ -57,6 +57,51 @@
...
+        "\nNF TAG         IID / SID / CORE    rx_pps  /  tx_pps        rx_drop  /  tx_drop"\
+        "         out   /    tonf     /   drop\n"\
+        "--------------------------------------------------------------------------------------------------------------------\n"
+#define ONVM_STATS_ADV_MSG \
+        "\nNF TAG         IID / SID / CORE    rx_pps  /  tx_pps        rx_drop  /  tx_drop"\
+        "           out   /    tonf     /   drop\n"\
+        "               PNT / S|W / CHLD  drop_pps  /  drop_pps      rx_drop  /  tx_drop"\
+        "           next  /    buf      /   ret\n"\
+        "----------------------------------------------------------------------------------------------------------------------\n"
+#define ONVM_STATS_SHARED_CPU_MSG \
```
Comment (koolzz) [inline onvm/onvm_mgr/onvm_stats.h]: How about we do it like this:
```
#define ONVM_STATS_SHARED_CPU_MSG "\n"\
        "NF TAG         IID / SID / CORE    rx_pps  /  tx_pps        rx_drop  /  tx_drop           out   /    tonf     /   drop\n"\
        "               PNT / S|W / CHLD  drop_pps  /  drop_pps      rx_drop  /  tx_drop           next  /    buf      /   ret\n"\
        "                                  wakeups  /  wakeup_rt\n"\
        "----------------------------------------------------------------------------------------------------------------------\n"
```

Even though the linter complains its a lot easier to read.

Comment (koolzz) [reply onvm/onvm_mgr/onvm_stats.h]: (applies to other defs)

Comment (kevindweb) [reply onvm/onvm_mgr/onvm_stats.h]: Sounds good just pushed that change

**Comment by koolzz** [comment]
@kevindweb fixed the conflicts, should be good to merge now

**Comment by kevindweb** [comment]
@onvm how many errors are here?

**Comment by onvm** [comment]
> @onvm how many errors are here?

## CI Message

Your results will arrive shortly

**Comment by koolzz** [comment]
@onvm .

**Comment by onvm** [comment]
> @onvm .

## CI Message

Your results will arrive shortly
